contracts/AirDrop.sol

```diff
@@ -0,0 +1,252 @@
+    /**
+     * @dev make ethers available for members to withdraw
+     *
+     * @param recepients_ - address of member
+     * @param amount_ - how many ethers will be available
+     * @param deadline_ - how long signature will be valid
+     * @param v_ - can be only 27 or 28
+     * @param r_ - contain info to recover signer
+     * @param s_ - contain info to recover signer
+     *
+     * NOTE : more information 'EIP712'
+     */
+    function dropEther(
+        address[] calldata recepients_,
+        uint256[] calldata amount_,
+        uint256 deadline_,
+        uint8 v_,
+        bytes32 r_,
+        bytes32 s_
+    ) external override onlyOwner {
+        require(
+            recepients_.length == amount_.length,
+            "Error : arrays have different length"
+        );
+        bytes32 structHash = keccak256(
+            abi.encode(
+                DROP_HASH,
+                keccak256(abi.encodePacked(amount_)),
+                deadline_,
+                keccak256(abi.encodePacked(recepients_)),
+                true
+            )
+        );
+        bytes32 digest = EIP712._hashTypedDataV4(structHash);
+
+        address msgSigner = ECDSA.recover(digest, v_, r_, s_);
+        require(msg.sender == msgSigner, "invalid signature");
+        require(
+            block.timestamp < deadline_,
+            "Error : signed transaction expired"
+        );
+        for (uint256 i; i < recepients_.length; i++) {
+            require(
+                amount_[i] != 0 && recepients_[i] != address(0),
+                "Error :'recepients_' or 'amount_' equal to 0"
+            );
+            balanceOfEther[recepients_[i]] += amount_[i];
+        }
+        emit DropEther(recepients_, amount_, deadline_);
+    }
+
+    /**
+     * @dev Change actually ERC20
+     *
+     * @param token_ - address of new ERC20 which connected
+     * to 'this' contract
+     */
+    function updateTokenAddress(address token_) external override onlyOwner {
+        require(
+            Address.isContract(token_) == true,
+            "Error : Incorrect address , only contract address"
+        );
+        token = IERC20(token_);
+        emit NewContractAddress(token_);
+    }
+
+    /**
+     * @dev This function return tokens to 'owner'
+     *
+     * NOTE: withdraw all available tokens
+     */
+    function withdrawTokens() external override onlyOwner {
+        uint256 balance = token.balanceOf(address(this));
+        require(balance > 0, "Error : No availabale tokens to withdraw");
+        token.safeTransfer(msg.sender, balance);
+        emit WithdrawTokens(msg.sender, balance);
+    }
+
+    /**
+     * @dev This function return ether to 'owner'
+     *
+     * NOTE: withdraw all available ether
+     */
+    function withdrawEther() external override onlyOwner {
+        require(
+            address(this).balance > 0,
+            "Error : No availabale ether to withdraw"
+        );
+        uint256 value = address(this).balance;
```

Comment:
move to 212 line before require, and use variable in require mor gas optimization